include/proxy-wasm/context.h

```diff
@@ -0,0 +1,350 @@
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace proxy_wasm {
+
+#include "proxy_wasm_common.h"
+#include "proxy_wasm_enums.h"
+
+class WasmBase;
+class WasmVm;
+
+using Pairs = std::vector<std::pair<string_view, string_view>>;
+using PairsWithStringValues = std::vector<std::pair<string_view, std::string>>;
+using CallOnThreadFunction = std::function<void(std::function<void()>)>;
+
+struct BufferInterface {
+  virtual ~BufferInterface() {}
+  virtual size_t size() const = 0;
+  // Returns true on success.
+  virtual bool copyTo(WasmBase *wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                      uint64_t size_ptr) const = 0;
+};
+
+// Opaque context object.
+class StorageObject {
+public:
+  virtual ~StorageObject() = default;
+};
+
+struct PluginBase {
+  PluginBase(string_view name, string_view root_id, string_view vm_id,
+             string_view plugin_configuration)
+      : name_(std::string(name)), root_id_(std::string(root_id)), vm_id_(std::string(vm_id)),
+        plugin_configuration_(plugin_configuration) {}
+
+  const std::string name_;
+  const std::string root_id_;
+  const std::string vm_id_;
+  std::string plugin_configuration_;
+  const std::string &log_prefix() const { return log_prefix_; }
+
+private:
+  std::string makeLogPrefix() const;
+
+  std::string log_prefix_;
+};
+
+// A context which will be the target of callbacks for a particular session
+// e.g. a handler of a stream.
+class ContextBase {
+public:
+  ContextBase();                                                   // Testing.
+  ContextBase(WasmBase *wasm);                                     // Vm Context.
+  ContextBase(WasmBase *wasm, std::shared_ptr<PluginBase> plugin); // Root Context.
+  ContextBase(WasmBase *wasm, uint32_t root_context_id,
+              std::shared_ptr<PluginBase> plugin); // Stream context.
+  virtual ~ContextBase();
+
+  WasmBase *wasm() const { return wasm_; }
+  uint32_t id() const { return id_; }
+  bool isVmContext() { return id_ == 0; }
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  // Called before deleting the context.
+  virtual void destroy();
+
+  //
+  // VM level downcalls into the WASM code on Context(id == 0).
+  //
+  virtual bool onStart(std::shared_ptr<PluginBase> plugin);
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  //
+  // Stream downcalls on Context(id > 0).
+  //
+  // General stream downcall on a new stream.
+  virtual void onCreate(uint32_t root_context_id);
+  // Network
+  virtual FilterStatus onNetworkNewConnection() {
```

Comment:
I have a high level question: It seems that this context object implements every possible method, even if that method is not used by a particular plugin. It seems strange to me that we would not have context objects that are either root/shared, and then derived classes that are specific to a type of plugin/context being used. Can you elaborate on the design decisions here? Also, presumably this is the ABI effectively? Are there parallel docs I should look at?

Reply:
There are several reasons for this design.   First,, it enables a single plugin to be used for multiple purposes.  For example, we register the Context supporting the Http filter for both Http and AccessLog callbacks.  However, one could also register a single plugin for both Http and Network if all you are doing is telemetry. 

Also, having all the callbacks on one Context object means that it implements the full ABI which is nice.  ,It also makes it easier to write tests since there is only one object which has to be overloaded within the test framework. It also makes the code simpler and succinct which itself makes it easier to read and maintain.  An inheritance hierarchy adds complexity and can lead to downcasting which is problematic.  Certainly separating things out does make it look simpler on when one is considering only a single class, but the interactions between the pieces are more complex, so that simplicity is IMO an illusion.

Reply:
> so that simplicity is IMO an illusion

I disagree, but as this is a matter of opinion, you wrote the code, and this is an extension / outside the repo, so be it.

However, I can't review this any further without looking at the ABI docs/definitions/etc. That is the most critical piece. Where is that? I don't think we can continue until that is done.

Reply:
https://github.com/proxy-wasm/spec is a version.   Perhaps we can get @PiotrSikora to do as I am doing with the host code and make a review PR.

Reply:
That would be really optimal. That is actually by far the most important thing that I think we need to get the most eyes on. Can we do that ASAP? cc @htuch

Reply:
@jplevyak one other comment/question/way forward. This file is basically the primary ABI, right? Shouldn't this file have pretty explicit function comments for each function? Maybe we can do that instead of the markdown since I think it would be better for the docs to live with the code and then we can just review here? WDYT?

Also, can you educate us on the ABI rules? Presumably, once a version is snapped, this file (and any associated C stubs) cannot be changed, right?

Reply:
Yes. I'll add documentation here.

Reply:
Added documentation comments PTAL

Reply:
RE: ABI rules.   I have added a VERSIONING.md file with a TODO to the repo.  We have not yet worked out all the details for versioning, however I do not think it will have a large effect on this file.  As long as ContextBase provides a superset of the required ABI behavior the actual ABI versioning will happen in wasm.{h,cc} and in the SDK.  We will be exporting different sets of calls in those files which will be implemented by the ContextBase API.

Reply:
Thanks, so just so I understand, the ABI is always C functions, right? So basically each of these functions has an associated C function which would map the context ID to the actual context and make the call? Is that right?

Reply:
Yep.